### PR TITLE
Adds support for deferrable exclude constraints in PostgreSQL.

### PR DESCRIPTION
**Pull request by alpaca-tc**
### Motivation / Background

By default, exclude constraints in PostgreSQL are checked after each row. This works for most use cases, but becomes a major limitation when replacing records with overlapping ranges by using multiple statements.

```ruby
exclusion_constraint :users, "daterange(valid_from, valid_to) WITH &&", deferrable: :immediate
```

Passing `deferrable: :immediate` checks constraint after each statement, but allows manually deferring the check using `SET CONSTRAINTS ALL DEFERRED` within a transaction. This will cause the excludes to be checked after the transaction.

It's also possible to change the default behavior from an immediate check (after the statement), to a deferred check (after the transaction):

```ruby
exclusion_constraint :users, "daterange(valid_from, valid_to) WITH &&", deferrable: :deferred
```

### Detail

This is follow up to #40224

### Additional information

It is with reference to [this comment](https://github.com/rails/rails/pull/46192#issuecomment-1438368442) that the `deferrable` option accepts only `:immediate` and `:deferred`.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb [hunk 1]
```diff
@@ -660,6 +664,8 @@ def unique_keys(table_name)
         # The +options+ hash can include the following keys:
         # [<tt>:name</tt>]
         #   The constraint name. Defaults to <tt>excl_rails_<identifier></tt>.
+        # [<tt>:deferrable</tt>]
+        #   Specify whether or not the unique constraint should be deferrable. Valid values are +false+ or +:immediate+ or +:deferred+ to specify the default behavior. Defaults to +false+.
```
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb:668]:
```suggestion
        #   Specify whether or not the exclusion constraint should be deferrable. Valid values are +false+ or +:immediate+ or +:deferred+ to specify the default behavior. Defaults to +false+.
```

File: activerecord/test/cases/migration/exclusion_constraint_test.rb [hunk 2]
```diff
@@ -74,6 +149,29 @@ def test_added_exclusion_constraint_ensures_valid_values
           end
         end
 
+        def test_added_deferrable_initially_immediate_exclusion_constraint
```
Comment (yahonda) [inline activerecord/test/cases/migration/exclusion_constraint_test.rb:152]: This test gets an error if it runs isolated.

```ruby
$ ARCONN=postgresql bin/test test/cases/migration/exclusion_constraint_test.rb -n test_added_deferrable_initially_immediate_exclusion_constraint
Using postgresql
Run options: -n test_added_deferrable_initially_immediate_exclusion_constraint --seed 54897

# Running:

E

Error:
ActiveRecord::Migration::ExclusionConstraintTest#test_added_deferrable_initially_immediate_exclusion_constraint:
NameError: uninitialized constant ActiveRecord::Migration::ExclusionConstraintTest::Section
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/migration/exclusion_constraint_test.rb:164:in `block in test_added_deferrable_initially_immediate_exclusion_constraint'
    /home/yahonda/src/github.com/rails/rails/activesupport/lib/active_support/testing/assertions.rb:48:in `assert_nothing_raised'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/migration/exclusion_constraint_test.rb:163:in `test_added_deferrable_initially_immediate_exclusion_constraint'


bin/test test/cases/migration/exclusion_constraint_test.rb:152



Finished in 0.025691s, 38.9244 runs/s, 38.9244 assertions/s.
1 runs, 1 assertions, 0 failures, 1 errors, 0 skips
$
```

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
Comment (yahonda) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:1]: This line is out of scope from this pull request, I've found the exclusion constraint name should be quoted. Would you include this change in this pull request?

```diff
$ git diff
diff --git a/activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb b/activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
index 9ec4dd4504..dbd19534f0 100644
--- a/activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
+++ b/activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb
@@ -37,7 +37,7 @@ def visit_ValidateConstraint(name)

           def visit_ExclusionConstraintDefinition(o)
             sql = ["CONSTRAINT"]
-            sql << o.name
+            sql << quote_column_name(o.name)
             sql << "EXCLUDE"
             sql << "USING #{o.using}" if o.using
             sql << "(#{o.expression})"
$
```

Comment (yahonda) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:1]: Ignoring this review comment should be fine if you do not prefer it.

**Review by yahonda — COMMENTED** [review]
Thanks for the follow-up pull request.

**Comment by alpaca-tc** [comment]
@yahonda I have fixed the PR based on your review comments. 😄

File: activerecord/test/cases/migration/exclusion_constraint_test.rb (hunk 2 above)
Comment (alpaca-tc) [reply activerecord/test/cases/migration/exclusion_constraint_test.rb:152]: Fixed to use `Invoice` instead of `Section`.

**Review by yahonda — APPROVED** [review]
